### PR TITLE
Improve lifetime management for PyProxy attributes

### PR DESCRIPTION
**Pull request by hoodmane**
This implements the features proposed in #1617.

### CODE REVIEW

File: src/js/pyodide.js [hunk 1]
```diff
@@ -235,11 +235,13 @@ def temp(Module):
...
-  Module.restoreState = (state) =>
-    Module.pyodide_py._state.restore_state(state);
-
   Module.init_dict.get("temp")(Module);
+
+  let _state = Module.pyodide_py._state;
+  let save_state = _state.save_state;
+  let restore_state = _state.restore_state;
+  Module.saveState = () => save_state();
+  Module.restoreState = (state) => restore_state(state);
```
Comment (rth) [inline src/js/pyodide.js:244]: We now have those in api.js I think? Syncing with main would probably show conflicts here.

File: src/core/pyproxy.js [hunk 2]
```diff
@@ -95,7 +96,7 @@ Module.disable_pyproxy_allocation_tracing();
  * still make "prototype in proxy" be true though.
  * @private
  */
-Module.pyproxy_new = function (ptrobj) {
+Module.pyproxy_new = function (ptrobj, cache) {
```
Comment (rth) [inline src/core/pyproxy.js:99]: Could you please mention cache in the above docstring.

File: src/core/pyproxy.js [hunk 3]
```diff
@@ -188,6 +194,46 @@ Module.getPyProxyClass = function (flags) {
...
 Module.PyProxy_getPtr = _getPtr;
+Module.pyproxy_mark_borrowed = function (proxy) {
+  proxy.$$.borrowed = true;
+};
+
+const pyproxy_cache_destroyed_msg =
+  "This borrowed attribute proxy was automatically destroyed in the " +
+  "process of destroying the proxy it was borrowed from. Try using the 'copy' method.";
+
+function pyproxy_destroy_cache(cache) {
```
Comment (rth) [inline src/core/pyproxy.js:205]: Maybe we could add a doc string with a sentence explaining what this function does. It's non obvious that despite its name it does not always destroy the cache (depending on the refcnt value). Maybe a better name would be `pyproxy_decref_cache` ?

**Review by rth — APPROVED** [review]
Thanks @hoodmane , aside for the following comments LGTM.

File: src/js/pyodide.js (hunk 1 above)
Comment (hoodmane) [reply src/js/pyodide.js:244]: Yeah they are better in a different file.

File: src/core/pyproxy.js (hunk 3 above)
Comment (hoodmane) [reply src/core/pyproxy.js:205]: Yeah I think I named it before I figured out that the cache would need to be refcounted.

File: src/core/pyproxy.js (hunk 2 above)
Comment (hoodmane) [reply src/core/pyproxy.js:99]: Also the existing docstring has a lot of unimportant information about cleaning up callable `PyProxy` objects which is duplicated in comments in the function body, so I'll edit that down.

**Comment by hoodmane** [comment]
Merging master caused a bunch of hiwire leak checks to fail again. Will have to investigate why.

**Comment by hoodmane** [comment]
I think now the only failing tests will be normal flakiness. Will merge if this CI run looks good.

**Pull request merged**